pkg/providers/instance/instance.go

```diff
@@ -138,36 +145,39 @@ func (p *DefaultProvider) Get(ctx context.Context, id string) (*Instance, error)
 
 func (p *DefaultProvider) List(ctx context.Context) ([]*Instance, error) {
 	var out = &ec2.DescribeInstancesOutput{}
-	err := p.ec2api.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
-		Filters: []*ec2.Filter{
+
+	paginator := ec2.NewDescribeInstancesPaginator(p.ec2api, lo.ToPtr(ec2.DescribeInstancesInput{
+		Filters: []ec2types.Filter{
 			{
 				Name:   aws.String("tag-key"),
-				Values: aws.StringSlice([]string{karpv1.NodePoolLabelKey}),
+				Values: []string{karpv1.NodePoolLabelKey},
 			},
 			{
 				Name:   aws.String("tag-key"),
-				Values: aws.StringSlice([]string{v1.LabelNodeClass}),
+				Values: []string{v1.LabelNodeClass},
 			},
 			{
 				Name:   aws.String("tag-key"),
-				Values: aws.StringSlice([]string{fmt.Sprintf("kubernetes.io/cluster/%s", options.FromContext(ctx).ClusterName)}),
+				Values: []string{fmt.Sprintf("kubernetes.io/cluster/%s", options.FromContext(ctx).ClusterName)},
 			},
 			instanceStateFilter,
 		},
-	}, func(page *ec2.DescribeInstancesOutput, _ bool) bool {
-		out.Reservations = append(out.Reservations, page.Reservations...)
-		return true
-	})
-	if err != nil {
-		return nil, fmt.Errorf("describing ec2 instances, %w", err)
+	}))
+
+	for paginator.HasMorePages() {
+		output, err := paginator.NextPage(ctx)
+		out.Reservations = append(out.Reservations, output.Reservations...)
```

Comment:
Where is `out` defined? I think you mean to get output here. This will never find any reservations

Reply:
Out is defined at the top of the function. it maintains the same functionality as the old function, we are getting the reservations from the current page and adding them to the overall list of reservations so after all pages are searched we will have all that exist to pass it through instancesFromOutput on line 174. I can change outputs naming to page to help with confusion.